### PR TITLE
✨  Adding OCP Deployment Support for KS Monitoring Tools

### PR DESCRIPTION
**Pull request by dumb0002**
## Summary

This PR adds support to deploy the KS monitoring tools in a OCP environment. 

## Related issue(s)

Fixes #

### CODE REVIEW

File: monitoring/grafana/grafana-route.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  to:
+    kind: Service
+    name: grafana
+    weight: 100
+  port:
+    targetPort: service
+  wildcardPolicy: None
+  tls:
+    termination: edge
+    insecureEdgeTerminationPolicy: Redirect
```
Comment (MikeSpreitzer) [inline monitoring/grafana/grafana-route.yaml:15]: Reviewers will see less red if this file has a terminal linebreak.

File: monitoring/install-ks-monitoring.sh [hunk 2]
```diff
@@ -21,10 +21,12 @@ ctx="kind-kubeflex"
 ns="ks-monitoring"
 opt="core" # (value: core or wec)
 hub_context="kind-kubeflex"
+env="k8s"
```
Comment (MikeSpreitzer) [inline monitoring/install-ks-monitoring.sh:24]: An OCP cluster _is_ a Kubernetes cluster. In other words, the category of "kubernetes" includes OCP as a subcategory. It would be clearer to use something more specific than "k8s" to contrast with "ocp". Perhaps "base" or "plain"?

File: monitoring/install-ks-monitoring.sh [hunk 3]
```diff
@@ -120,29 +163,45 @@ if [ $opt == "core" ];then
...
 
+    if [ $env == "ocp" ];then
+         kubectl -n $ns apply -f ${SCRIPT_DIR}/grafana/grafana-route.yaml
+    fi 
+
     # Set the Thanos URL for Prometheus remote write
     export thanos_host="thanos-receive.ks-monitoring.svc.cluster.local:19291"
-    sed -e s/%THANOS_HOST%/$thanos_host/g ${SCRIPT_DIR}/prometheus/prometheus-custom-values.yaml >& /tmp/prometheus-custom-values-set.yaml
+    endpoint="http://$thanos_host"
+    sed -e s,%THANOS_HOST%,$endpoint,g -e s,%ENABLE_KUBEAPISERVER%,true,g ${SCRIPT_DIR}/prometheus/prometheus-custom-values.yaml >& /tmp/prometheus-custom-values-set.yaml
```
Comment (MikeSpreitzer) [inline monitoring/install-ks-monitoring.sh:173]: (A) See https://www.gnu.org/software/bash/manual/bash.html#Redirecting-Standard-Output-and-Standard-Error for the preferred syntax for redirecting both stdout and stderr.

(B) Why redirect stderr? You do not want error messages in that output YAML file, right?

File: monitoring/install-ks-monitoring.sh [hunk 4]
```diff
@@ -175,4 +234,4 @@ wait-for-object $ctx $ns statefulset pyroscope
 wait-for-object $ctx $ns statefulset pyroscope-alloy
 
 # Delete temporary file
-rm /tmp/pyroscope-config-values-set.yaml
\ No newline at end of file
+rm /tmp/pyroscope-config-values-set.yaml
```
Comment (MikeSpreitzer) [inline monitoring/install-ks-monitoring.sh:237]: There is more to remove now, right?

File: monitoring/README.md [hunk 5]
```diff
@@ -107,7 +111,7 @@ wds1-ks-transport-controller   64m
 b) ITS space:
 
 ```bash
-./configure-monitoring-its.sh 
+./configure-monitoring-its.sh --host-cluster-context kind-kubeflex
 ```
```
Comment (MikeSpreitzer) [inline monitoring/README.md:116]: BTW, the set of Grafana dashboards in the picture at new line 174 is outdated; there I see 3, but now there are 4.

```console
mspreitz@mjs13 monitoring % ls grafana/*.json
grafana/APIServerMonitoring.json			grafana/KubeStellarControllersMonitoring-Average.json
grafana/APIServerPriorityFairness.json			grafana/KubeStellarControllersMonitoring.json
```

File: monitoring/grafana/grafana-route.yaml (hunk 1 above)
Comment (dumb0002) [reply monitoring/grafana/grafana-route.yaml]: Fixed!

File: monitoring/install-ks-monitoring.sh (hunk 2 above)
Comment (dumb0002) [reply monitoring/install-ks-monitoring.sh:24]: IMO, the terms "k8s" and "ocp" envs or clusters are very popular and less ambiguous than using very generic terms like "base" or "plain" - what are your thoughts on this?

File: monitoring/install-ks-monitoring.sh (hunk 4 above)
Comment (dumb0002) [reply monitoring/install-ks-monitoring.sh:237]: Could you provide more details? I don't understand this comment. At each stage of deployment of the components "temporary files" are being deleted.

File: monitoring/prometheus/thanos-custom-values.yaml [hunk 6]
```diff
@@ -34,9 +34,9 @@ receive:
...
   ingress:
-    enabled: true
+    #enabled: true
     hostname: kubeflex-control-plane
     ingressClassName: nginx
   persistence:
     enabled: true
-    size: 1Gi
\ No newline at end of file
+    size: 2Gi
```
[the remark this answers is not in the document]
Comment (dumb0002) [reply monitoring/prometheus/thanos-custom-values.yaml]: Fixed!

File: monitoring/prometheus/thanos-route.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  to:
+    kind: Service
+    name: thanos-receive
+    weight: 100
+  port:
+    targetPort: remote
+  wildcardPolicy: None
+  tls:
+    termination: edge
+    insecureEdgeTerminationPolicy: Redirect
```
[the remark this answers is not in the document]
Comment (dumb0002) [reply monitoring/prometheus/thanos-route.yaml]: Fixed!

File: monitoring/README.md [hunk 8]
```diff
@@ -27,13 +27,15 @@ This example shows how to deploy monitoring tools (Prometheus/Thanos, Grafana an
```
[the remark this answers is not in the document]
Comment (dumb0002) [reply monitoring/README.md:27]: Fixed! I update OCM status addon chart to version `v0.2.0-rc14`

File: monitoring/README.md (hunk 5 above)
Comment (dumb0002) [reply monitoring/README.md:116]: Yes, the newly added Grafana dashboard will be useful for debugging for me it some situations but I don't want to expose it to the end-user. Hence, intentionally, I did not update the documentation to reference this new Grafana dashboard.

File: monitoring/install-ks-monitoring.sh (hunk 3 above)
Comment (dumb0002) [reply monitoring/install-ks-monitoring.sh]: Good catch! I update the script to redirect the stdout only.

File: monitoring/install-ks-monitoring.sh (hunk 2 above)
Comment (MikeSpreitzer) [reply monitoring/install-ks-monitoring.sh:24]: First, let's talk about what `--help` produces. What you have got there is in the syntax of one terse line, but the content has a lot of explanatory material. I think that a better model for the syntax for the output from `--help` would be what you get from, for example, `make --help`:

```console
mspreitz@mjs13 kubestellar % make --help
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
...
```

With this format, you have room for explanatory material.

You can explain that "--env plain" means "plain kubernetes".

File: monitoring/README.md (hunk 5 above)
Comment (MikeSpreitzer) [reply monitoring/README.md:116]: But you have two references, one is updated and one is not. One refers the reader to everything in the directory, the other shows a snapshot of the result of loading those.

Comment (dumb0002) [reply monitoring/README.md:116]: Ok, I see! To avoid any further confusion, I removed the file `KubeStellarControllersMonitoring-Average.json`

File: monitoring/install-ks-monitoring.sh (hunk 2 above)
Comment (dumb0002) [reply monitoring/install-ks-monitoring.sh]: Ok, I see! I replaced `k8s` with `plain` as you suggested above, and it's the default value.

File: monitoring/install-ks-monitoring.sh [hunk 9]
```diff
@@ -21,10 +21,12 @@ ctx="kind-kubeflex"
 ns="ks-monitoring"
 opt="core" # (value: core or wec)
 hub_context="kind-kubeflex"
+env="plain"
+
 
 while [ $# != 0 ]; do
     case "$1" in
-        (-h|--help) echo "$0 usage: (--cluster-context (e.g., --cluster-context core-cluster (default value: kind-kubeflex)) | --monitoring-ns (e.g., --monitoring-ns monitoring (default value: ks-monitoring)) | --set (e.g., --set wec (default value: core)))*"
+        (-h|--help) echo "$0 usage: (--cluster-context (e.g., --cluster-context core-cluster (default value: kind-kubeflex)) | --monitoring-ns (e.g., --monitoring-ns monitoring (default value: ks-monitoring)) | --set (e.g., --set wec (default value: core)) | --env (e.g., --env ocp (default value: plain)) | --hub-context (e.g., --hub-context kscore (default value: kind-kubeflex)) *"
```
Comment (MikeSpreitzer) [inline monitoring/install-ks-monitoring.sh:29]: For the non-obvious flags, the help should list all the allowed values rather than just one example.

I think it would be better to follow the common format for a help message that includes explanations. Following is an example.

```console
mspreitz@mjs13 kubestellar % make --help
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
...
```

File: monitoring/README.md [hunk 10]
```diff
@@ -58,12 +60,14 @@ thanos-storegateway-0                                  1/1     Running   0
 ```
 
 
-b) Deploy the monitoring tools in a WEC cluster (e.g., `cluster1`):
+b) Deploy the monitoring tools in a plain K8s WEC cluster (e.g., `cluster1`):
```
Comment (MikeSpreitzer) [inline monitoring/README.md:63]: I see `hub_context` being used in the plain kube case too, on lines 199 and 200 of the current edition of the script. They read as follows.

```shell
  export ROOT_USER=$(kubectl --context $hub_context get secret -n $ns minio -o jsonpath="{.data.rootUser}" | base64 -d)
  export ROOT_PASSWORD=$(kubectl --context $hub_context get secret -n $ns minio -o jsonpath="{.data.rootPassword}" | base64 -d)
```

File: monitoring/README.md [hunk 11]
```diff
@@ -86,7 +90,7 @@ pyroscope-alloy-0                                      2/2     Running   0
 a) WDS space:
 
 ```bash
-./configure-monitoring-wds.sh
+./configure-monitoring-wds.sh --host-cluster-context kind-kubeflex
```
Comment (MikeSpreitzer) [inline monitoring/README.md:93]: `--host-cluster-context` here means the same thing as `--hub-context` for the other script, right? If so then I recommend picking one and sticking with it. In KubeStellar we tend to talk about the "KubeFlex hosting cluster", while OCM/ACM talks about a "hub cluster" --- which corresponds to an ITS in KubeStellar.

File: monitoring/README.md [hunk 12]
```diff
@@ -58,12 +60,14 @@ thanos-storegateway-0                                  1/1     Running   0
...
 
 
-b) Deploy the monitoring tools in a WEC cluster (e.g., `cluster1`):
+b) Deploy the monitoring tools in a plain K8s WEC cluster (e.g., `cluster1`):
 
 ```bash
 ./install-ks-monitoring.sh --cluster-context cluster1 --set wec
 ```
 
+For OpenShift clusters add the flags `--env ocp` & `--hub-context <hub-context-name>` in the commad above.
```
Comment (MikeSpreitzer) [inline monitoring/README.md:69]: In KubeStellar terms, is this a context for the KubeFlex hosting cluster or the ITS?

**Comment by MikeSpreitzer** [comment]
Not a blocker for this PR, since this problem was not introduced here, but it is getting more important. There should be documentation about the configuration produced. There is a picture, but no more. For someone who will want to debug anything going wrong, a bit more description of the configuration would be helpful. Stuff like Service objects, port numbers, protocols, HTTP vs HTTPS, who/what is authorized to read, write.

File: monitoring/prometheus/minio-route.yaml [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  to:
+    kind: Service
+    name: minio-console
+    weight: 100
+  port:
+    targetPort: http
+  wildcardPolicy: None
+  tls:
+    termination: edge
+    insecureEdgeTerminationPolicy: Redirect
```
Comment (MikeSpreitzer) [inline monitoring/prometheus/minio-route.yaml:31]: FYI, no final linebreak in this file.

Comment (dumb0002) [reply monitoring/prometheus/minio-route.yaml]: Fixed!

File: monitoring/README.md (hunk 12 above)
Comment (dumb0002) [reply monitoring/README.md:69]: The flag `--cluster-context`  refers to either the context of the hosting KubeFlex cluster OR of a WEC cluster - depending when the monitoring tooling is being deployed.

File: monitoring/README.md (hunk 11 above)
Comment (dumb0002) [reply monitoring/README.md:93]: Yes!

File: monitoring/install-ks-monitoring.sh (hunk 9 above)
Comment (dumb0002) [reply monitoring/install-ks-monitoring.sh]: I added the list of allowed values for the `--env` flag. 
I kept consistency of the help command output with the others KS scripts. I can update it in a new PR as it would require updating many other scripts to keep the consistency of the output of the 'help' command.

Comment (MikeSpreitzer) [reply monitoring/install-ks-monitoring.sh]: I do not think that we need this sort of consistency. Unix/Linux CLI users are already coping with a diversity of styles of usage documentation. My point was that this particular script seems suited to the multi-line style. OK to address this in follow-on.

File: monitoring/README.md (hunk 11 above)
Comment (MikeSpreitzer) [reply monitoring/README.md]: In the current rev of this PR, I only see "--hub-context" appearing in one place in this file, new line 69. So I am not understanding what you updated to match what.

I find the usage of the term "hub" in the context of KubeStellar to be confusing. In KubeStellar we use "core" to refer to all the central stuff and "KubeFlex hosting cluster" to refer to that specific thing. In plain English it would be reasonable to use "hub" to mean "the central stuff", but that positively conflicts with the technical term "hub" from OCM, which is important because KubeStellar uses OCM and a KubeStellar ITS is an OCM hub. So i see two plausible meanings of the word "hub" and I am rarely sure what anyone means by it. If you mean specifically the KubeFlex hosting cluster, then I strongly recommend using words like that. And where you mean a KubeStellar ITS, I strongly recommend using words like that.

File: monitoring/README.md (hunk 12 above)
Comment (MikeSpreitzer) [reply monitoring/README.md:69]: Sorry, I was not clear. My question is not about `--cluster-context`. My question is about `--hub-context`. Does this mean the KubeFlex hosting cluster or does it mean a KubeStellar ITS?

File: monitoring/README.md (hunk 10 above)
Comment (MikeSpreitzer) [reply monitoring/README.md:63]: My point here is that the instructions here do NOT say to provide `--hub-context` in the plain kube case, but it seems like `--hub-context` is needed in the plain kube case too.

File: monitoring/install-ks-monitoring.sh (hunk 9 above)
Comment (dumb0002) [reply monitoring/install-ks-monitoring.sh]: Sounds good, I will address it in a separate PR.

File: monitoring/README.md (hunk 10 above)
Comment (dumb0002) [reply monitoring/README.md:63]: The `--hub-context` flag has a default value that is used for the plain kube case. That's why that flag is omitted for the kube case instructions.

File: monitoring/README.md (hunk 12 above)
Comment (dumb0002) [reply monitoring/README.md:69]: `--hub-context` means the KubeFlex hosting cluster.

File: monitoring/README.md [hunk 14]
```diff
@@ -107,7 +111,7 @@ wds1-ks-transport-controller   64m
 b) ITS space:
 
 ```bash
-./configure-monitoring-its.sh
```
[the remark this answers is not in the document]
Comment (dumb0002) [reply monitoring/README.md:110]: The flag `--host-context` refers to the KubeFlex hosting cluster only. Hence, for the script **configure-monitoring-wec.sh** we're using this flag instead: `--wec-cluster-context`

File: monitoring/README.md (hunk 11 above)
Comment (dumb0002) [reply monitoring/README.md]: The flag `--hub-context` refers to the KubeFlex hosting cluster OR any other cluster that is hosting the core components of the KS control plane.

File: monitoring/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply monitoring/README.md:110]: Could configure-monitoring-wds.sh involve only one context, the one for accessing the WDS?

Could configure-monitoring-its.sh involve only one context, the one for accessing the ITS?

It would be easier to understand if each of these three scripts only involves one context, for accessing the one space being configured.

File: monitoring/README.md (hunk 10 above)
Comment (MikeSpreitzer) [reply monitoring/README.md:63]: That default value is "kind-kubeflex", right? Does the `monitoring/install-ks-monitoring.sh` script only work for OCP and the kind-based demo environment? If it works for cases where the KubeFlex hosting cluster is plain kube but not created by `kind` in our usual way then this should be documented. And if it is so restricted, that should also be documented.

File: monitoring/README.md (hunk 11 above)
Comment (MikeSpreitzer) [reply monitoring/README.md]: What would be an example of an "other cluster that is hosting the core components of the KS control plane" but is not the KubeFlex hosting cluster?

File: monitoring/README.md (hunk 10 above)
Comment (dumb0002) [reply monitoring/README.md:63]: Ok, I see! The script `monitoring/install-ks-monitoring.sh` is not restrictive. It works too for any plain k8s environment besides kind. To make this clear, I will add the in the instructions for the plain kube case too.

File: monitoring/README.md (hunk 11 above)
Comment (dumb0002) [reply monitoring/README.md]: If I am not mistaken, there were some discussion about not having a tight coupling between KubeFlex and KS and to allow deployment of KS control plane in other cluster. Maybe these are just future plans. Then, I will replace the flag name `--hub-context` with `kubeflex-hosting-cluster-context`.

File: monitoring/README.md (hunk 12 above)
Comment (dumb0002) [reply monitoring/README.md:69]: FYI, for better clarity, I replaced the flag name `--hub-context` with `kubeflex-hosting-cluster-context`.

File: monitoring/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (dumb0002) [reply monitoring/README.md:110]: For better clarity, I just updated one of the flags for these scripts. Hence, both script have the following flags:
   a) `--kubeflex-hosting-cluster-context`: context of the KubeFlex hosting cluster;
   b) `--space-name`:  KS space name, for example: either ws1 and its1.

File: monitoring/README.md (hunk 11 above)
Comment (MikeSpreitzer) [reply monitoring/README.md]: If I understand correctly, this script only functions correctly if the context in question accesses a KubeFlex hosting cluster. For example, the use of `$wds-system` is specific to KubeFlex hosting. Unless you have a vision of how this script would work correctly in the general case, I recommend having the name of this flag be specific to KubeFlex hosting.

File: monitoring/README.md [hunk 15]
```diff
@@ -86,7 +90,7 @@ pyroscope-alloy-0                                      2/2     Running   0
 a) WDS space:
 
 ```bash
-./configure-monitoring-wds.sh
+./configure-monitoring-wds.sh --kubeflex-hosting-cluster-context kind-kubeflex
 ```
 
 If you have more than one WDS, you can run the above script with the `--space-name` flag. For example: `./configure-monitoring-wds.sh --space-name wds2`
```
Comment (MikeSpreitzer) [inline monitoring/README.md:96]: I think this PR did not introduce this problem, but I encountered it while testing. The problem is the assumption that a space from a KubeFlex ControlPlane named "foo" is accessed through a kubeconfig context named "foo". That is a convenient pattern that we generally follow, but is not a requirement elsewhere in KubeStellar. Now that I am working with three different systems (two triples of clusters in the IBM cloud, plus local testing with a triple of `kind` clusters), I can not use the one context name "wds1" for three different contexts. I am currently testing with a system where I use a context named "dev-wds1" to access the space defined by the ControlPlane named "wds1". Following is a failure.

```console
bash-3.2$ time ./configure-monitoring-wds.sh --kubeflex-hosting-cluster-context $host_context --space-name dev-wds1
+ set -e
+ ctx=kind-kubeflex
+ wds=wds1
+ monitoring_ns=ks-monitoring
+ '[' 4 '!=' 0 ']'
+ case "$1" in
+ ((  4 > 1  ))
+ ctx=default/c106-e-us-south-containers-cloud-ibm-com:31891/IAM#mspreitz@us.ibm.com
+ shift
+ shift
+ '[' 2 '!=' 0 ']'
+ case "$1" in
+ ((  2 > 1  ))
+ wds=dev-wds1
+ shift
+ shift
+ '[' 0 '!=' 0 ']'
+++ dirname -- ./configure-monitoring-wds.sh
++ cd -- .
++ pwd
+ SCRIPT_DIR=/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring
+ kubectl config use-context default/c106-e-us-south-containers-cloud-ibm-com:31891/IAM#mspreitz@us.ibm.com
Switched to context "default/c106-e-us-south-containers-cloud-ibm-com:31891/IAM#mspreitz@us.ibm.com".
+ : --------------------------------------------------------------------
+ : Configure APF for prometheus traffic to the WDS API server
+ : --------------------------------------------------------------------
+ kubectl --context dev-wds1 apply -f /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/configuration/prometheus-pod-exempt.yaml
Warning: flowcontrol.apiserver.k8s.io/v1beta3 FlowSchema is deprecated in v1.29+, unavailable in v1.32+; use flowcontrol.apiserver.k8s.io/v1 FlowSchema
flowschema.flowcontrol.apiserver.k8s.io/prometheus-exempt created
+ : --------------------------------------------------------------------
+ : Configure kubestellar controller manager pod for prometheus scraping
+ : --------------------------------------------------------------------
+ : 1. Create service monitor for KS ctl manager
+ sed s/%WDS%/dev-wds1/g /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/configuration/ks-ctl-manager-sm.yaml
+ kubectl -n ks-monitoring apply -f -
servicemonitor.monitoring.coreos.com/dev-wds1-ks-controller-manager created
+ : --------------------------------------------------------------------
+ : Configure kubestellar controller manager pod for pyroscope scraping
+ : --------------------------------------------------------------------
+ : 1. Add labels for pyroscope scraping:
+ kubectl -n dev-wds1-system get deploy kubestellar-controller-manager -o yaml
+ yq '.spec.template.metadata.annotations."profiles.grafana.com/cpu.port" |= "8082" |  .spec.template.metadata.annotations."profiles.grafana.com/cpu.scrape"|= "true" | .spec.template.metadata.annotations."profiles.grafana.com/goroutine.port" |= "8082" | .spec.template.metadata.annotations."profiles.grafana.com/goroutine.scrape" |= "true" |
.spec.template.metadata.annotations."profiles.grafana.com/memory.port" |= "8082" | .spec.template.metadata.annotations."profiles.grafana.com/memory.scrape" |= "true"'
+ kubectl -n dev-wds1-system apply -f -
Error from server (NotFound): namespaces "dev-wds1-system" not found
error: error validating "STDIN": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false

real	0m2.812s
user	0m0.391s
sys	0m0.207s
```

File: monitoring/README.md [hunk 16]
```diff
@@ -27,13 +27,15 @@ This example shows how to deploy monitoring tools (Prometheus/Thanos, Grafana an
 
 Starting from a local directory containing the git repo, do the following:
 
-a) Deploy the monitoring tools in the hosting cluster (e.g., `kind-kubeflex`):
+a) Deploy the monitoring tools in a plain K8s hosting cluster (e.g., `kind-kubeflex`):
 
 ```bash
 cd monitoring
 ./install-ks-monitoring.sh --cluster-context kind-kubeflex --set core
```
Comment (MikeSpreitzer) [inline monitoring/README.md:34]: FYI, I tested this and was surprised at how long it took.

```console
bash-3.2$ time ./install-ks-monitoring.sh --cluster-context $host_context --set core --env ocp
+ set -e
...
All replicas are ready.
+ break
+ rm /tmp/pyroscope-config-values-set.yaml

real	7m19.363s
user	0m29.312s
sys	0m11.035s
```

**Comment by MikeSpreitzer** [comment]
Using kubeconfig context names established by `oc login` to IBM Cloud OCP clusters, which include a slash, gets a failure.

```console
bash-3.2$ echo $wec1_context
default/c107-e-us-south-containers-cloud-ibm-com:30728/IAM#mspreitz@us.ibm.com

bash-3.2$ ./configure-monitoring-wec.sh --wec-cluster-context $wec1_context
+ set -e
+ ctx=cluster1
+ monitoring_ns=ks-monitoring
+ '[' 2 '!=' 0 ']'
+ case "$1" in
+ ((  2 > 1  ))
+ ctx=default/c107-e-us-south-containers-cloud-ibm-com:30728/IAM#mspreitz@us.ibm.com
+ shift
+ shift
+ '[' 0 '!=' 0 ']'
+++ dirname -- ./configure-monitoring-wec.sh
++ cd -- .
++ pwd
+ SCRIPT_DIR=/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring
+ kubectl config use-context default/c107-e-us-south-containers-cloud-ibm-com:30728/IAM#mspreitz@us.ibm.com
Switched to context "default/c107-e-us-south-containers-cloud-ibm-com:30728/IAM#mspreitz@us.ibm.com".
+ : --------------------------------------------------------------------
+ : Configure Status Agent-Addon pod for Prometheus scraping
+ : --------------------------------------------------------------------
+ : 1. Create status agent-addon service:
+ kubectl -n open-cluster-management-agent-addon apply -f /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/configuration/status-agent-svc.yaml
service/status-agent-controller created
+ : 2. Create service monitor for status agent-addon
+ sed s/%WEC%/default/c107-e-us-south-containers-cloud-ibm-com:30728/IAM#mspreitz@us.ibm.com/g /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/monitoring/configuration/status-agent-sm.yaml
+ kubectl -n ks-monitoring apply -f -
sed: 1: "s/%WEC%/default/c107-e- ...": bad flag in substitute command: 'c'
error: no objects passed to apply
```

**Comment by MikeSpreitzer** [comment]
This is a little worrying.

image

**Comment by MikeSpreitzer** [comment]
The hosting cluster and the WECs have the same IP address for the apiserver, in my testing.

```console
mspreitz@mjs13 ~ % kubectl --context $host_context get Endpoints -A | egrep "NAME| wds| vcluster |kubernetes"
NAMESPACE                                          NAME                                                              ENDPOINTS                                                                 AGE
default                                            kubernetes                                                        172.20.0.1:2040                                                           114d
its1-system                                        vcluster                                                          172.30.2.57:8443,172.30.2.57:8443                                         11h
wds1-system                                        wds1                                                              172.30.172.85:9444                                                        11h
mspreitz@mjs13 ~ % 
mspreitz@mjs13 ~ % 
mspreitz@mjs13 ~ % kubectl --context wec1 get Endpoints -n default 
NAME                        ENDPOINTS            AGE
kubernetes                  172.20.0.1:2040      93d
openshift-apiserver         172.19.193.160:443   93d
openshift-oauth-apiserver   172.19.59.187:443    93d
mspreitz@mjs13 ~ % 
mspreitz@mjs13 ~ % kubectl --context wec2 get Endpoints -n default 
NAME                        ENDPOINTS            AGE
kubernetes                  172.20.0.1:2040      114d
openshift-apiserver         172.19.207.112:443   114d
openshift-oauth-apiserver   172.19.242.106:443   114d
```

File: monitoring/README.md [hunk 17]
```diff
@@ -153,6 +157,7 @@ a) Port-forward Grafana to localhost, by using the kubectl command:
 ```bash
 kubectl --context kind-kubeflex port-forward -n ks-monitoring svc/grafana 3000:80
 ```
+For OpenShift clusters get the OpenShift route for the Grafana service in the `ks-monitoring` namespace.
```
Comment (MikeSpreitzer) [inline monitoring/README.md:160]: I do not understand this sentence. What exactly is it telling me to get (an API object, a URL, ...?), and what am I supposed to do with it?

File: monitoring/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply monitoring/README.md:110]: This particular thread is not about the naming of flags but the design of what flags are needed. The ITS and WDS script are given the hosting cluster context and the ControlPlane name, while the WEC script is only given the context for the WEC. I am asking if we can simplify the ITS and WDS scripts to that they too are just given one contex (the ITS and WDS space contexts, respectively)?

**Review by MikeSpreitzer — APPROVED** [review]
I think this still needs work but it can be done in follow-on.

/lgtm
/approve
Thanks!
